pkg/util/util.go

```diff
@@ -173,3 +175,72 @@ func GetDefaultGatewayInterface() (*net.Interface, error) {
 
 	return nil, fmt.Errorf("unable to find default route")
 }
+
+func CreateChainIfNotExists(ipt *iptables.IPTables, table, chain string) error {
+	existingChains, err := ipt.ListChains(table)
+	if err != nil {
+		return err
+	}
+
+	for _, val := range existingChains {
+		if val == chain {
+			// Chain already exists
+			return nil
+		}
+	}
+
+	return ipt.NewChain(table, chain)
+}
+
+func PrependUnique(ipt *iptables.IPTables, table string, chain string, ruleSpec []string) error {
+	rules, err := ipt.List(table, chain)
+	if err != nil {
+		return fmt.Errorf("error listing the rules in %s chain: %v", chain, err)
+	}
+
+	// Submariner requires certain iptable rules to be programmed at the beginning of an iptables Chain
+	// so that we can preserve the sourceIP for inter-cluster traffic and avoid K8s SDN making changes
+	// to the traffic.
+	// In this API, we check if the required iptable rule is present at the beginning of the chain.
+	// If the rule is already present and there are no stale[1] flows, we simply return. If not, we create one.
+	// [1] Sometimes after we program the rule at the beginning of the chain, K8s SDN might insert some
+	// new rules ahead of the rule that we programmed. In such cases, the rule that we programmed will
+	// not be the first rule to hit and Submariner behavior might get affected. So, we query the rules
+	// in the chain to see if the rule slipped its position, and if so, delete all such occurrences.
+	// We then re-program a new rule at the beginning of the chain as required.
+
+	isPresentAtRequiredPosition := false
+	numOccurrences := 0
+	for index, rule := range rules {
+		if strings.Contains(rule, strings.Join(ruleSpec, " ")) {
+			klog.V(4).Infof("In %s table, iptables rule \"%s\", exists at index %d.", table, strings.Join(ruleSpec, " "), index)
+			numOccurrences++
+
+			if index == 1 {
+				isPresentAtRequiredPosition = true
+			}
+		}
+	}
+
+	// The required rule is present in the Chain, but either there are multiple occurrences or its
+	// not at the desired location
+	if numOccurrences > 1 || !isPresentAtRequiredPosition {
+		for i := 0; i < numOccurrences; i++ {
+			if err = ipt.Delete(table, chain, ruleSpec...); err != nil {
+				return fmt.Errorf("error deleting stale iptable rule \"%s\": %v", strings.Join(ruleSpec, " "), err)
+			}
+		}
+	}
+
+	// The required rule is present only once and is at the desired location
+	if numOccurrences == 1 && isPresentAtRequiredPosition {
+		klog.V(4).Infof("In %s table, iptables rule \"%s\", already exists.", table, strings.Join(ruleSpec, " "))
+		return nil
+	} else {
+		if err = ipt.Insert(table, chain, 1, ruleSpec...); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
```

Comment:
Based on best practices outlined [here](https://github.com/golang/go/wiki/CodeReviewComments#package-names), I suggest putting these functions in a separate _iptables_ package and file.

Reply:
We are already using an external iptables package (github.com/coreos/go-iptables/iptables). These APIs are more like utility APIs that internally use the actual iptables package. Creating an iptables package with these APIs IMO might be confusing and shadowing the actual iptables package. WDYS?

Reply:
I understand. Not a big deal - it's fine.